storybookjs/storybook
# CLI: Automatically detect typescript in `sb init`

Issue: #5050 

## What I did
This PR adds automatic detection for typescript projects when installing storybook by running `npx @storybook/cli sb init`.
With that, it's not necessary to pass `--story-format` anymore (unless it is desired)!

While I was at it, I revamped the framework detection code to be more scalable. If you need to add a new template, all you have to do is create a new entry at `supportedTemplates` in `project_types.js` like the following:
```js
{
    preset: projectTypes.NEW_SUPPORTED_TEMPLATE,
    dependencies: [], // <-- optional, tests for these both as dependencies and devDependencies
    peerDependencies: [], // <-- optional
    files: [], // <-- optional
    matcherFunction: ({ dependencies, files, peerDependencies }) => {
      // every argument is returned as an array of booleans.
      return // whatever assertion you want, as long as it returns boolean.
    },
}
```
As for the typescript detection heuristics, I believe the common point in typescript projects is `typescript` as dependency. We can definitely change it to use more assertions, such as checking if files like `tsconfig.json` or `tslint.json` exist, or if the project contains at least one `*.ts|*.tsx` file.

I tried really hard to test and assert that the implementation works well, and comparing with the current implementation (in next), this is the only difference in the output (before/after):
<img src="https://user-images.githubusercontent.com/1671563/80286269-a97ae080-872a-11ea-8890-07a1d92c0a58.png" width="600">

But of course there could be edge cases or something I missed, so please help me out make sure this is good. Also please be thorough in the review, spare no comments! even with the simplest thing. 👍 In case you want to discuss, we can have a call if that makes it easier!

## How to test
Checkout this branch, run `yarn build cli`, go to `libs/cli` and run `yarn test`. Tweak with the test script if you want/ You can also check and run the unit tests `helpers.test.js` and `detect.test.js`.

## Does this need an update to the documentation?
I guess [this page](https://storybook.js.org/docs/configurations/typescript-config/) could be updated so the init command is mentioned there, however the page is about general typescript configuration, and currently storybook doesn't have templates for every framework + ts.

## Next steps
* Turn the files to typescript, so there's no way to make silly mistakes/typo and so that the code is easier to understand and use. I added comments but types would have been much better. I didn't convert to TS in this PR because it means turning a lot of other files to typescript as well, and this PR would be even bigger. I'm willing to do that though in a separate PR!
* Add more typescript templates. Currently there is only for react (not counting for angular, which is ts by default). I'm also willing to add templates for the remaining ones.
* Add support for monorepos? Maybe for Lerna based monorepos, as NX for instance has a different way of setting up storybook, using `@nrwl/storybook`.
